api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,645 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+ ****************************************************************************
+\page page_scatter_gather_emulation Emulating x86 Scatter and Gather Instructions
+
+\tableofcontents
+
+# Background
+
+The x86 gather and scatter instructions were introduced in the AVX2 and AVX512
+instruction set extensions. They allow loading or storing a subset of elements in a
+vector from/to multiple non-contiguous addresses.
+
+AVX2 has only gather instructions, no scatter instructions, whereas AVX512 has
+both. AVX2 is limited to 256-bit length vectors, whereas AVX512 has 512-bit
+support. Both support masking of individual memory accesses, using either a special
+mask register in AVX512 or another vector in AVX2.
+
+Examples of these instructions are (in DR’s IR):
+```
+vpgatherdd %rax(,%ymm11,4)[4byte] %ymm13 -> %ymm12 %ymm13
+```
+
+Above is an AVX2 gather instruction that reads 32-bit doublewords into the 256-bit
+`ymm12` vector from addresses generated by adding the base address in `rax` to the
+corresponding index elements in `ymm11`, conditionally based on the masks in
+`ymm13`. Elements may be gathered in any order. When an element is read, its mask
+is cleared. If some load faults, all elements to its right (closer to LSB) will
+be complete.
+
+
+```
+vpscatterdd {%k1} %xmm10 -> %rax(,%xmm11,4)[4byte] %k1
+```
+
+Above is an AVX512 scatter instruction that writes 32-bit doublewords from the 128-
+bit `xmm10` vector to addresses generated by adding the base address in `rax` to
+the corresponding index element in `xmm11`, conditionally based on the mask
+register `k1`. Elements may be scattered in any order. When an element is stored,
+its mask is cleared. If some store faults, all elements to its right (closer to
+LSB) will be complete.
+
+
+# Problem Statement
+
+Scatter and gather instructions pose a challenge to DynamoRIO clients that observe
+memory addresses, like address tracing tools (e.g. drcachesim, that collects
+memory address and control flow traces), and taint tracking tools. They are complex
+to handle because:
+
+-  A single gather or scatter instruction loads from or stores to multiple
+addresses
+-  Accessed addresses may be non-contiguous
+-  Each access is conditional based on some mask
+
+DynamoRIO clients only see the scatter or gather instruction and need to do more
+work to extract all accessed addresses. This is unlike regular scalar loads or
+stores, where the accessed address is readily available. The goal of this work is
+to make it easier for DR clients to observe these addresses. We achieve this by
+expanding the scatter and gather instructions into a functionally equivalent
+sequence of scalar stores and loads. This way, DR clients will see regular store
+and load instructions which they can instrument as usual. This is similar to what
+DR does for repeat string operations (like `rep`, `repnz`): convert it into a
```

Comment:
A complete opcode like `rep movs` seems clearer than just the prefix.

Reply:
Fixed.